src/kriscv/sparse_bytes.py

```diff
@@ -0,0 +1,191 @@
+from __future__ import annotations
+
+import copy
+from dataclasses import dataclass
+from itertools import pairwise
+
+from pyk.kast.inner import KInner
+from pyk.prelude.kint import eqInt, intToken
+from pyk.prelude.ml import mlEqualsTrue
+
+from .term_builder import (
+    add_bytes,
+    bytesToken,
+    dot_sb,
+    length_bytes,
+    normalize_memory,
+    sb_bytes,
+    sb_bytes_cons,
+    sb_empty,
+    sb_empty_cons,
+)
+
+
+def _size(data: bytes | int | SymBytes) -> int:
+    if isinstance(data, bytes):
+        return len(data)
+    elif isinstance(data, int):
+        return data
+    else:
+        return data.size
+
+
+def _split(
+    data: bytes | int | SymBytes, offset: int
+) -> tuple[bytes | int | SymBytes | None, bytes | int | SymBytes | None]:
+    size = _size(data)
+    if offset == 0:
+        return None, data
+    elif offset == size:
+        return data, None
+    elif 0 < offset < size:
+        if isinstance(data, bytes):
+            return data[:offset], data[offset:]
+        elif isinstance(data, int):
+            return offset, size - offset
+        elif isinstance(data, SymBytes):
+            raise NotImplementedError('Splitting symbolic bytes is not implemented')
+        else:
+            raise ValueError(f'Unexpected item type: {type(data)}')
+    else:
+        raise ValueError(f'Offset {offset} is out of bounds for size {size}')
+
+
+@dataclass(frozen=True)
+class SymBytes:
+    data: KInner
+    size: int
+
+    def constraint(self) -> KInner:
+        return mlEqualsTrue(eqInt(length_bytes(self.data), intToken(self.size)))
+
+
+@dataclass
+class SparseBytes:
```

Comment:
```suggestion
class SparseBytes(Mapping[int, SparseBytes]):
```

Does this type check?

Reply:
'SparseBytes' is undefined in `Mapping[int, SparseBytes]`. Actually, even for `Mapping[int, bytes | int | SymBytes]`, we cannot define `_iter_` for it. The correct structure for that might be: `Mapping[tuple(int, int), bytes| Symbytes]`. But I don't think that we need `_iter_` for now.